File: substrate/highlight_source_attach_quality_interrogation_twin_compose.py
"""Highlight source attach quality interrogation → twin feed (pure).

live_dispatched / merge_executed / twin_written / remote_fetched always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.highlight_source_attach_quality_interrogation_compose import (
    HighlightSourceAttachQualityInterrogationCompose,
    HighlightSourceAttachQualityInterrogationComposeError,
    compose_highlight_source_attach_quality_interrogation,
)
from substrate.twin_chase_analysis_feed_compose import (
    TwinChaseAnalysisFeedCompose,
    TwinChaseAnalysisFeedComposeError,
    compose_twin_chase_analysis_feed,
)


class HighlightSourceAttachQualityInterrogationTwinComposeError(ValueError):
    """Fail-closed validation for highlight source twin pack."""


@dataclass(frozen=True)
class HighlightSourceAttachQualityInterrogationTwinCompose:
    parent_asset_id: str
    session_id: str
    highlight_pack: HighlightSourceAttachQualityInterrogationCompose
    twin_feed: TwinChaseAnalysisFeedCompose
    pack_ready: bool
    live_dispatched: bool
    merge_executed: bool
    remote_fetched: bool
    twin_written: bool
    prompts_injected: bool
    store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "parent_asset_id": self.parent_asset_id,
            "session_id": self.session_id,
            "highlight_pack": self.highlight_pack.to_dict(),
            "twin_feed": self.twin_feed.to_dict(),
            "pack_ready": self.pack_ready,
            "live_dispatched": False,
            "merge_executed": False,
            "remote_fetched": False,
            "twin_written": False,
            "prompts_injected": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "authority": (
                "highlight_source_attach_quality_interrogation_twin_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise HighlightSourceAttachQualityInterrogationTwinComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _derive_findings(
    highlight: str,
    sources: object,
    questions: object,
) -> list[dict[str, Any]]:
    if not isinstance(sources, list):
        raise HighlightSourceAttachQualityInterrogationTwinComposeError(
            "sources must be an array"
        )
    if not isinstance(questions, list):
        raise HighlightSourceAttachQualityInterrogationTwinComposeError(
            "questions must be an array"
        )
    findings: list[dict[str, Any]] = [
        {"source_id": "hl-highlight", "body": highlight, "kind": "data"}
    ]
    for s in sources:
        if not isinstance(s, dict):
            raise HighlightSourceAttachQualityInterrogationTwinComposeError(
                "sources items must be objects"
            )
        sid = str(s.get("source_id", "")).strip()
        title = str(s.get("title", "")).strip()
        if not sid or not title:
            raise HighlightSourceAttachQualityInterrogationTwinComposeError(
                "sources items need source_id and title"
            )
        findings.append(
            {"source_id": f"src-{sid}", "body": title, "kind": "data"}
        )
    for q in questions:
        if not isinstance(q, dict):
            raise HighlightSourceAttachQualityInterrogationTwinComposeError(
                "questions items must be objects"
            )
        qid = str(q.get("question_id", "")).strip()
        body = str(q.get("body", "")).strip()
        if not qid or not body:
            raise HighlightSourceAttachQualityInterrogationTwinComposeError(
                "questions items need question_id and body"
            )
        findings.append(
            {"source_id": f"q-{qid}", "body": body, "kind": "question"}
        )
    return findings


def compose_highlight_source_attach_quality_interrogation_twin(
    *,
    parent_asset_id: object,
    highlight: object,
    gated: object,
    would_exceed: object,
    operator_ack: object,
    session_id: object,
    requested_families: object,
    sources: object,
    quality_overall: object,
    questions: object,
    chase_mode: object,
    models: object,
    daily_cap_usd: object,
    spent_usd: object,
    prompt: object | None = None,
    preferred_view_mode: object | None = None,
    operator_override: object | None = None,
    selected_model_id: object | None = None,
    source_families: object | None = None,
    citations: object | None = None,
    derive_citations_from_sources: object | None = None,
    quality_floor: object | None = None,
    prior_records: object | None = None,
    user_prompt: object | None = None,
    projected_cost_usd_high: object | None = None,
    projected_cost_usd_low: object | None = None,
    bench_bests: object | None = None,
    focus_task: object | None = None,
    nd_shadow: object | None = None,
    require_both: object | None = None,
    existing_twin_asset_id: object | None = None,
    analysis_excerpt: object | None = None,
    mark_for_prompt_context: object | None = None,
    twin_findings: object | None = None,
    require_both_with_twin: object | None = None,
) -> HighlightSourceAttachQualityInterrogationTwinCompose:
    """Highlight pack + twin feed. Never dispatches/writes twin."""
    if not isinstance(operator_ack, bool):
        raise HighlightSourceAttachQualityInterrogationTwinComposeError(
            "operator_ack must be an explicit boolean"
        )
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")
    session = _require_nonempty(session_id, field="session_id")
    hl = _require_nonempty(highlight, field="highlight")

    require_twin = (
        True if require_both_with_twin is None else require_both_with_twin
    )
    if not isinstance(require_twin, bool):
        raise HighlightSourceAttachQualityInterrogationTwinComposeError(
            "require_both_with_twin must be boolean when set"
        )

    mark_prompt = (
        True if mark_for_prompt_context is None else mark_for_prompt_context
    )
    if not isinstance(mark_prompt, bool):
        raise HighlightSourceAttachQualityInterrogationTwinComposeError(
            "mark_for_prompt_context must be boolean when set"
        )

    notes: list[str] = [
        "live_dispatched=false · merge_executed=false",
        "remote_fetched=false · twin_written=false · prompts_injected=false",
    ]

    try:
        highlight_pack = compose_highlight_source_attach_quality_interrogation(
            parent_asset_id=parent,
            highlight=hl,
            gated=gated,
            would_exceed=would_exceed,
            operator_ack=operator_ack,
            session_id=session,
            requested_families=requested_families,
            sources=sources,
            quality_overall=quality_overall,
            questions=questions,
            chase_mode=chase_mode,
            models=models,
            daily_cap_usd=daily_cap_usd,
            spent_usd=spent_usd,
            prompt=prompt,
            preferred_view_mode=preferred_view_mode,
            operator_override=operator_override,
            selected_model_id=selected_model_id,
            source_families=source_families,
            citations=citations,
            derive_citations_from_sources=derive_citations_from_sources,
            quality_floor=quality_floor,
            prior_records=prior_records,
            user_prompt=user_prompt,
            projected_cost_usd_high=projected_cost_usd_high,
            projected_cost_usd_low=projected_cost_usd_low,
            bench_bests=bench_bests,
            focus_task=focus_task,
            nd_shadow=nd_shadow,
            require_both=require_both,
        )
    except HighlightSourceAttachQualityInterrogationComposeError as e:
        raise HighlightSourceAttachQualityInterrogationTwinComposeError(
            str(e)
        ) from e
    notes.extend(f"[highlight_pack] {n}" for n in highlight_pack.notes)

    if twin_findings is not None:
        if not isinstance(twin_findings, list):
            raise HighlightSourceAttachQualityInterrogationTwinComposeError(
                "twin_findings must be an array when set"
            )
        findings = twin_findings
        notes.append(f"twin_findings={len(findings)} caller-supplied")
    else:
        findings = _derive_findings(hl, sources, questions)
        notes.append(
            f"twin_findings={len(findings)} derived from "
            "highlight+sources+questions"
        )

    try:
        twin_feed = compose_twin_chase_analysis_feed(
            session_id=session,
            parent_asset_id=parent,
            findings=findings,
            operator_ack=operator_ack,
            analysis_excerpt=analysis_excerpt,
            existing_twin_asset_id=existing_twin_asset_id,
            mark_for_prompt_context=mark_prompt,
        )
    except TwinChaseAnalysisFeedComposeError as e:
        raise HighlightSourceAttachQualityInterrogationTwinComposeError(
            str(e)
        ) from e
    notes.extend(f"[twin_feed] {n}" for n in twin_feed.notes)

    if require_twin:
        pack_ready = (
            highlight_pack.pack_ready is True
            and twin_feed.feed_ready is True
            and operator_ack is True
        )
    else:
        pack_ready = operator_ack is True and (
            highlight_pack.pack_ready is True or twin_feed.feed_ready is True
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — highlight pack + twin feed ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — highlight pack, twin feed, or operator_ack "
            "gate open"
        )

    if (
        highlight_pack.live_dispatched is not False
        or highlight_pack.merge_executed is not False
        or twin_feed.twin_written is not False
        or twin_feed.prompts_injected is not False
    ):
        raise HighlightSourceAttachQualityInterrogationTwinComposeError(
            "invariant: honesty flags must remain false"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "merge_executed=false",
            "remote_fetched=false",
            "twin_written=false",
            "prompts_injected=false",
            "store_mutated=false",
        )
    )

    return HighlightSourceAttachQualityInterrogationTwinCompose(
        parent_asset_id=parent,
        session_id=session,
        highlight_pack=highlight_pack,
        twin_feed=twin_feed,
        pack_ready=pack_ready,
        live_dispatched=False,
        merge_executed=False,
        remote_fetched=False,
        twin_written=False,
        prompts_injected=False,
        store_mutated=False,
        notes=tuple(notes),
        authority=(
            "highlight_source_attach_quality_interrogation_twin_compose_advisory"
        ),
    )


def format_highlight_source_attach_quality_interrogation_twin_summary(
    c: HighlightSourceAttachQualityInterrogationTwinCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"highlight_ready={c.highlight_pack.pack_ready} · "
        f"twin_feed_ready={c.twin_feed.feed_ready} · "
        f"findings={c.twin_feed.finding_count} · "
        f"live_dispatched=false · twin_written=false · merge_executed=false"
    )


__all__ = [
    "HighlightSourceAttachQualityInterrogationTwinCompose",
    "HighlightSourceAttachQualityInterrogationTwinComposeError",
    "compose_highlight_source_attach_quality_interrogation_twin",
    "format_highlight_source_attach_quality_interrogation_twin_summary",
]

File: apps/reading/src/api/highlightSourceAttachQualityInterrogationTwinCompose.ts
/**
 * Highlight source attach quality interrogation → twin feed (pure).
 *
 * Operator vision: from a highlight, attach sources, chase questions, feed
 * recursive twin note-taker so reading insights inform future work.
 *
 * live_dispatched / merge_executed / twin_written / remote_fetched always false.
 */

import {
  composeHighlightSourceAttachQualityInterrogation,
  type HighlightSourceAttachQualityInterrogationCompose,
  type HighlightSourceAttachQualityInterrogationInput,
} from "./highlightSourceAttachQualityInterrogationCompose";
import {
  composeTwinChaseAnalysisFeed,
  type ChaseFeedFinding,
  type TwinChaseAnalysisFeedCompose,
} from "./twinChaseAnalysisFeedCompose";

export interface HighlightSourceAttachQualityInterrogationTwinInput
  extends HighlightSourceAttachQualityInterrogationInput {
  existing_twin_asset_id?: string | null;
  analysis_excerpt?: string | null;
  mark_for_prompt_context?: boolean;
  twin_findings?: ChaseFeedFinding[] | null;
  require_both_with_twin?: boolean;
}

export interface HighlightSourceAttachQualityInterrogationTwinCompose {
  parent_asset_id: string;
  session_id: string;
  highlight_pack: HighlightSourceAttachQualityInterrogationCompose;
  twin_feed: TwinChaseAnalysisFeedCompose;
  pack_ready: boolean;
  live_dispatched: false;
  merge_executed: false;
  remote_fetched: false;
  twin_written: false;
  prompts_injected: false;
  store_mutated: false;
  notes: string[];
  authority: "highlight_source_attach_quality_interrogation_twin_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

function deriveFindings(
  highlight: string,
  sources: HighlightSourceAttachQualityInterrogationInput["sources"],
  questions: HighlightSourceAttachQualityInterrogationInput["questions"],
): ChaseFeedFinding[] {
  const findings: ChaseFeedFinding[] = [
    { source_id: "hl-highlight", body: highlight, kind: "data" },
  ];
  for (const s of sources) {
    findings.push({
      source_id: `src-${s.source_id}`,
      body: s.title,
      kind: "data",
    });
  }
  for (const q of questions) {
    findings.push({
      source_id: `q-${q.question_id}`,
      body: q.body,
      kind: "question",
    });
  }
  return findings;
}

export function composeHighlightSourceAttachQualityInterrogationTwin(
  input: HighlightSourceAttachQualityInterrogationTwinInput,
): HighlightSourceAttachQualityInterrogationTwinCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const highlight = requireNonEmpty(input.highlight, "highlight");

  const require_both_with_twin =
    input.require_both_with_twin === undefined
      ? true
      : input.require_both_with_twin;
  if (typeof require_both_with_twin !== "boolean") {
    throw new Error("require_both_with_twin must be boolean when set");
  }

  const mark_for_prompt_context =
    input.mark_for_prompt_context === undefined
      ? true
      : input.mark_for_prompt_context;
  if (typeof mark_for_prompt_context !== "boolean") {
    throw new Error("mark_for_prompt_context must be boolean when set");
  }

  const notes: string[] = [
    "live_dispatched=false · merge_executed=false",
    "remote_fetched=false · twin_written=false · prompts_injected=false",
  ];

  const highlight_pack = composeHighlightSourceAttachQualityInterrogation({
    parent_asset_id: input.parent_asset_id,
    highlight: input.highlight,
    gated: input.gated,
    prompt: input.prompt,
    preferred_view_mode: input.preferred_view_mode,
    would_exceed: input.would_exceed,
    operator_override: input.operator_override,
    selected_model_id: input.selected_model_id,
    source_families: input.source_families,
    operator_ack: input.operator_ack,
    session_id: input.session_id,
    requested_families: input.requested_families,
    sources: input.sources,
    citations: input.citations,
    derive_citations_from_sources: input.derive_citations_from_sources,
    quality_overall: input.quality_overall,
    quality_floor: input.quality_floor,
    questions: input.questions,
    chase_mode: input.chase_mode,
    prior_records: input.prior_records,
    user_prompt: input.user_prompt,
    models: input.models,
    daily_cap_usd: input.daily_cap_usd,
    spent_usd: input.spent_usd,
    projected_cost_usd_high: input.projected_cost_usd_high,
    projected_cost_usd_low: input.projected_cost_usd_low,
    bench_bests: input.bench_bests,
    focus_task: input.focus_task,
    nd_shadow: input.nd_shadow,
    require_both: input.require_both,
  });
  notes.push(...highlight_pack.notes.map((n) => `[highlight_pack] ${n}`));

  let twin_findings: ChaseFeedFinding[];
  if (input.twin_findings != null) {
    if (!Array.isArray(input.twin_findings)) {
      throw new Error("twin_findings must be an array when set");
    }
    twin_findings = input.twin_findings;
    notes.push(`twin_findings=${twin_findings.length} caller-supplied`);
  } else {
    twin_findings = deriveFindings(highlight, input.sources, input.questions);
    notes.push(
      `twin_findings=${twin_findings.length} derived from highlight+sources+questions`,
    );
  }

  const twin_feed = composeTwinChaseAnalysisFeed({
    session_id,
    parent_asset_id,
    findings: twin_findings,
    analysis_excerpt: input.analysis_excerpt,
    existing_twin_asset_id: input.existing_twin_asset_id,
    operator_ack: input.operator_ack,
    mark_for_prompt_context,
  });
  notes.push(...twin_feed.notes.map((n) => `[twin_feed] ${n}`));

  let pack_ready = false;
  if (require_both_with_twin) {
    pack_ready =
      highlight_pack.pack_ready === true &&
      twin_feed.feed_ready === true &&
      input.operator_ack === true;
  } else {
    pack_ready =
      input.operator_ack === true &&
      (highlight_pack.pack_ready === true || twin_feed.feed_ready === true);
  }

  if (pack_ready) {
    notes.push("pack_ready=true — highlight pack + twin feed ready; still pure");
  } else {
    notes.push(
      "pack_ready=false — highlight pack, twin feed, or operator_ack gate open",
    );
  }

  if (
    highlight_pack.live_dispatched !== false ||
    highlight_pack.merge_executed !== false ||
    twin_feed.twin_written !== false ||
    twin_feed.prompts_injected !== false
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("live_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("remote_fetched=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("store_mutated=false");

  return {
    parent_asset_id,
    session_id,
    highlight_pack,
    twin_feed,
    pack_ready,
    live_dispatched: false,
    merge_executed: false,
    remote_fetched: false,
    twin_written: false,
    prompts_injected: false,
    store_mutated: false,
    notes,
    authority:
      "highlight_source_attach_quality_interrogation_twin_compose_advisory",
  };
}

export function formatHighlightSourceAttachQualityInterrogationTwinSummary(
  c: HighlightSourceAttachQualityInterrogationTwinCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `highlight_ready=${c.highlight_pack.pack_ready} · ` +
    `twin_feed_ready=${c.twin_feed.feed_ready} · ` +
    `findings=${c.twin_feed.finding_count} · ` +
    `live_dispatched=false · twin_written=false · merge_executed=false`
  );
}
